validation: Stricter assumeutxo error handling when renaming chainstates

ryanofsky opened the pull request:
There are two places in assumeutxo code where it is calling `AbortNode` to trigger asynchronous shutdowns without returning errors to calling functions.

One case, in `LoadChainstate`, happens when snapshot validation succeeds, and there is an error trying to replace the background chainstate with the snapshot chainstate.

The other case, in `InvalidateCoinsDBOnDisk`, happens when snapshot validatiion fails, and there is an error trying to remove the snapshot chainstate.

In both cases the node is being forced to shut down, so it makes sense for these functions to raise errors so callers can know that an error happened without having to infer it from the shutdown state.

Noticed these cases while reviewing #27861, which replaces the `AbortNode` function with a `FatalError` function.

ryanofsky committed:
validation: Stricter assumeutxo error handling in InvalidateCoinsDBOnDisk

Currently InvalidateCoinsDBOnDisk is calling AbortNode without an error to the
caller if it fails. Change it to return just return util::Result, and update
the caller to handle the error itself.

This causes the secondary error to be shown below the main error instead of the
other way around.

ryanofsky committed:
validation: Stricter assumeutxo error handling in LoadChainstate

Make LoadChainstate return an explicit error when snapshot validation succeeds,
but there is an error trying to replace the background chainstate with the
snapshot chainstate. Previously in this case LoadChainstate would trigger a
shutdown and return INTERRUPTED, now it will return an actual error code.

There's no real change to behavior other than error message being formatted a
little differently.

Motivation for this change is to replace error handling via callbacks with
error handling via return value ahead of
https://github.com/bitcoin/bitcoin/pull/27861

TheCharlatan commented on src/node/chainstate.cpp:
```diff
@@ -207,7 +207,9 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
-            AbortNode("Background chainstate cleanup failed unexpectedly.");
+            const auto error = Untranslated("Background chainstate cleanup failed unexpectedly.");
+            AbortNode(error.original);
+            return {ChainstateLoadStatus::FAILURE, error};
```
Just an observation on the behavior change here. In the current behavior, the function continues execution until it reaches `CompleteChainstateInitialization` further below. There it calls `options.check_interrupt()`, which will lead it to return `ChainstateLoadStatus::INTERRUPTED`. I think making it return a `ChainstateLoadStatus::FAILURE` now makes more sense, since it is not stopped due to an external interrupt.

TheCharlatan commented on src/validation.cpp:
```diff
@@ -5661,6 +5661,7 @@ void Chainstate::InvalidateCoinsDBOnDisk()
             "snapshot directory %s, otherwise you will encounter the same error again "
             "on the next startup.",
             src_str, dest_str, src_str));
+        throw;
```
This seems like a significant behavior change. The way I read this, `InvalidateCoinsDBOnDisk` is only called by `MaybeCompleteSnapshotValidation`. It's return type is ignored by its call site in `ConnectTip`, which I think would now lead us to skip over some code afterwards.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/27862#discussion_r1227196719

> This seems like a significant behavior change.

It does only happen when snapshot validation fails (so the node would be shutting down anyway), and renaming the snapshot also fails. Just reraising the exception seemed like this simplest possible approach to take (copied from code in `ValidatedSnapshotCleanup`). But it is simple enough to clean up the code more. I pushed a new version which gets rid of the AbortNode call and uses `util::Result` instead.

TheCharlatan replied:
Nice, this is exactly what I had in mind! Can be resolved.

ryanofsky reviewed (commented):
Updated 7564b87ebab27bea76efe5dd54d3f6f41cc78a1b -> 1ac09b93cdb41eb7dbc1a62364363e59507da1af ([`pr/assumeabort.1`](https://github.com/ryanofsky/bitcoin/commits/pr/assumeabort.1) -> [`pr/assumeabort.2`](https://github.com/ryanofsky/bitcoin/commits/pr/assumeabort.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/assumeabort.1..pr/assumeabort.2)) splitting this into two commits and making more changes to `InvalidateCoinsDBOnDisk` to normalize its error handling

sedited reviewed (approved):
ACK 1ac09b93cdb41eb7dbc1a62364363e59507da1af

That windows build failure is weird though. How can there be a conflict between `std::format` and `tfm::format`, if we are only in the latter's namespace? Why is this only popping up now?

hebasto commented:
> No idea if #27892 fixes the compile error, but the best way to find out would be to rebase this pull on top of that?

It [works](https://cirrus-ci.com/task/5585639102480384).

hebasto commented:
> For ref, the msvc error is:

The other possible solution could be as follows:
```diff
--- a/src/util/translation.h
+++ b/src/util/translation.h
@@ -63,8 +63,8 @@ inline T const& TranslateArg(const T& arg, bool translated)
 template <typename... Args>
 bilingual_str format(const bilingual_str& fmt, const Args&... args)
 {
-    return bilingual_str{format(fmt.original, TranslateArg(args, false)...),
-                         format(fmt.translated, TranslateArg(args, true)...)};
+    return bilingual_str{tinyformat::format(fmt.original, TranslateArg(args, false)...),
+                         tinyformat::format(fmt.translated, TranslateArg(args, true)...)};
 }
 } // namespace tinyformat
 
```

ryanofsky reviewed (commented):
Updated 1ac09b93cdb41eb7dbc1a62364363e59507da1af -> 3984b7c14d5d1ade485c4b30c74641c9361f4c1d ([`pr/assumeabort.2`](https://github.com/ryanofsky/bitcoin/commits/pr/assumeabort.2) -> [`pr/assumeabort.3`](https://github.com/ryanofsky/bitcoin/commits/pr/assumeabort.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/assumeabort.2..pr/assumeabort.3)) pulling in #27892 to try to fix the MSVC error, and dropping the AbortNode call in the first commit to make the shutdown sequence more straightforward

TheCharlatan commented on src/node/chainstate.cpp:
```diff
@@ -207,7 +207,7 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
         LogPrintf("[snapshot] cleaning up unneeded background chainstate, then reinitializing\n");
         if (!chainman.ValidatedSnapshotCleanup()) {
-            AbortNode("Background chainstate cleanup failed unexpectedly.");
+            return {ChainstateLoadStatus::FAILURE, Untranslated("Background chainstate cleanup failed unexpectedly.")};
```
Should this be `FAILURE_FATAL` now?

ryanofsky commented:
Rebased 3984b7c14d5d1ade485c4b30c74641c9361f4c1d -> 1c7d08b9acd33aff343228ada7e058e606cb1062 ([`pr/assumeabort.3`](https://github.com/ryanofsky/bitcoin/commits/pr/assumeabort.3) -> [`pr/assumeabort.4`](https://github.com/ryanofsky/bitcoin/commits/pr/assumeabort.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/assumeabort.3-rebase..pr/assumeabort.4)) on top of #27892 and switched to ChainstateLoadStatus::FAILURE_FATAL in the first commit to avoid triggering reindex suggestion

sedited reviewed (approved):
ACK 1c7d08b9acd33aff343228ada7e058e606cb1062

achow101 commented:
ACK 1c7d08b9acd33aff343228ada7e058e606cb1062

As someone slightly unfamiliar with init and shutdown, `AbortNode` not actually killing the node when it is called is surprising to me. The changes in this PR make these snapshot failure cases line up better with what I had expected.

jamesob reviewed (approved):
ACK 1c7d08b9acd33aff343228ada7e058e606cb1062 ([`jamesob/ackr/27862.1.ryanofsky.validation_stricter_assu`](https://github.com/jamesob/bitcoin/tree/ackr/27862.1.ryanofsky.validation_stricter_assu))

Good, common sense changes. Thanks for this cleanup. Reviewed and built each commit locally.
